native_client/deepspeech.h

```diff
@@ -48,6 +48,16 @@ typedef struct Metadata {
   double confidence;
 } Metadata;
 
+/**
+ * @brief Stores Metadata structs for each alternative transcription
+ */
+typedef struct Result {
```

Comment:
Do we really need this in fact? We could just return a `Metadata*` and the caller already gave us the number of expected maximul alternatives, so he knows how much he needs / how to browse the list.

Reply:
There are cases where we don't have as many beams to return as the caller has asked for. For example when the user requests more results than `beam_width`. Also if not enough beams have been expanded (for example on a very short stream). We could use an invasive linked list approach but I'm not sure if that's worth the trouble given the added difficulty of writing the bindings with a linked list.

Reply:
Sure, but then we just return a smaller list than expected, and the caller should anyway check for null pointers, so we would just `NULL` remainders. Are there any edge case I might be missing ?

Reply:
Rationale here is also that we don't need to change all the bindings, while exposing a struct is much more work to update all the bindings, so we should do it if it's really needed :)

Reply:
What do you mean `NULL` remainders? This is a contiguous list, there's no other pointers to `NULL` other than the root. I guess we could null out the `items` pointer inside the remaining `Metadata` values and set `num_items` = 0.

Reply:
But that would require the caller to adhere to this custom protocol of checking if `result[i]->items` is null to find the end of the sequence.

Reply:
No, just:
```
Metadata** DS_SpeechToTextWithMetadata();
```

And then
```
int maxResults = 8;
Metadata** ms = malloc(sizeof(Metadata*)*8);
ms = DS_SpeechToTextWithMetadata(..., maxResults);
for (int i = 0; i < maxResults; i++) {
  if (!ms) {
    // end of the list before maxResult
    break;
  }
  // do something
}
```

Reply:
That makes sense.

Reply:
(But I think this would still require updating all the bindings.)

Reply:
Yes, but the update would be much simpler. As much as I can tell, it's not a weird enough way that a lot of other project do that. The first example I have in mind is within linux drivers, like:
```
/* table of devices that work with this driver */
static const struct usb_device_id skel_table[] = {
	{ USB_DEVICE(USB_SKEL_VENDOR_ID, USB_SKEL_PRODUCT_ID) },
	{ }					/* Terminating entry */
};
```
And so code iterating checks for nullity. Would there be any other justification for keeping this struct ? Again, my mind could miss something.

Reply:
Only for avoiding the extra indirection I guess. In the Linux example, what is the definition of `USB_DEVICE`? That's also a list of objects rather than pointers so I don't understand what "nullity" means.

Reply:
From what I can tell in the MODULE_DEVICE_TABLE example the sizes are always static and then a script computes the number of elements by doing `sizeof(table)/sizeof(usb_device_id)`: https://github.com/torvalds/linux/blob/master/scripts/mod/file2alias.c#L338

But our sizes are dynamic so we'd need the extra indirection.